backend/models/user.js

```diff
@@ -74,12 +91,57 @@ const findOne = async function (query, projection) {
 	return await DB.findOne("admin", COLL_NAME, query, projection);
 };
 
+const handleAuthenticateFail = async function (username) {
+	const currentTime = Date.now();
+
+	try {
+		const user = await User.findByUserName(username);
+		const elapsedTime = user.customData.lastFailedLoginAt ?
+			currentTime - user.customData.lastFailedLoginAt : undefined;
+
+		const failedLoginCount = user.customData.failedLoginCount && elapsedTime && elapsedTime < config.lockoutDuration ?
+			user.customData.failedLoginCount + 1 : 1;
+
+		const loginLocked = failedLoginCount >= config.maxUnsuccessfulLoginAttempts ? true : undefined;
+
+		await User.update(username, {
+			"customData.lastFailedLoginAt": currentTime,
+			"customData.failedLoginCount": failedLoginCount,
+			"customData.loginLocked": loginLocked
+		});
+
+		return Math.max(config.maxUnsuccessfulLoginAttempts - failedLoginCount, 0);
+	} catch(err) {
+		// suppress update failure
```

Comment:
What kind of failures are we suppressing here?

Reply:
@carmenfan If the user doesn't exist it goes through here as well. I guess I should change that and maybe pass in the already retrieve `user` as well.